yjit_codegen.c

```diff
@@ -2898,6 +2898,127 @@ gen_send(jitstate_t *jit, ctx_t *ctx)
+        // In the interpreter the method id can change which is tested for and
+        // invalidates the cache.
+        // By skipping super calls inside a BMETHOD definition, I believe we
+        // avoid this case
+        return YJIT_CANT_COMPILE;
+    }
+
+    VALUE current_defined_class = me->defined_class;
+    ID mid = me->def->original_id;
+
+    // vm_search_normal_superclass
+    if (BUILTIN_TYPE(current_defined_class) == T_ICLASS && FL_TEST_RAW(RBASIC(current_defined_class)->klass, RMODULE_IS_REFINEMENT)) {
+        return YJIT_CANT_COMPILE;
+    }
+    VALUE comptime_superclass = RCLASS_SUPER(RCLASS_ORIGIN(current_defined_class));
+
+    const struct rb_callinfo *ci = cd->ci;
+    int32_t argc = (int32_t)vm_ci_argc(ci);
+
+    // Don't JIT calls that aren't simple
+    // Note, not using VM_CALL_ARGS_SIMPLE because sometimes we pass a block.
+    if ((vm_ci_flag(ci) & (VM_CALL_KW_SPLAT | VM_CALL_KWARG | VM_CALL_ARGS_SPLAT | VM_CALL_ARGS_BLOCKARG)) != 0) {
+        GEN_COUNTER_INC(cb, send_callsite_not_simple);
+        return YJIT_CANT_COMPILE;
+    }
+
+    VALUE comptime_recv = jit_peek_at_stack(jit, ctx, argc);
+    VALUE comptime_recv_klass = CLASS_OF(comptime_recv);
+
+    // Ensure we haven't rebound this method onto an incompatible class.
+    // In the interpreter we try to avoid making this check by performing some
+    // cheaper calculations first, but since we specialize on the receiver
+    // class and so only have to do this once at compile time this is fine to
+    // always check and side exit.
+    if (!rb_obj_is_kind_of(comptime_recv, current_defined_class)) {
+        return YJIT_CANT_COMPILE;
+    }
+
+    // Because we're assuming only one current_defined_class for a given
+    // receiver class we need to check that the superclass doesn't also
+    // re-include the same module.
+    if (rb_class_search_ancestor(comptime_superclass, current_defined_class)) {
+        return YJIT_CANT_COMPILE;
+    }
+
+    // Do method lookup
+    const rb_callable_method_entry_t *cme = rb_callable_method_entry(comptime_superclass, mid);
+
+    if (!cme) {
+        return YJIT_CANT_COMPILE;
+    }
+
+    // Check that we'll be able to write this method dispatch before generating checks
+    switch (cme->def->type) {
+        case VM_METHOD_TYPE_ISEQ:
+        case VM_METHOD_TYPE_CFUNC:
+            break;
+        default:
+            // others unimplemented
+            return YJIT_CANT_COMPILE;
+    }
+
+    // Guard that the receiver has the same class as the one from compile time
+    uint8_t *side_exit = yjit_side_exit(jit, ctx);
+
+    if (!block) {
+        // Guard no block passed
+        // rb_vm_frame_block_handler(GET_EC()->cfp) == VM_BLOCK_HANDLER_NONE
+        // note, we assume VM_ASSERT(VM_ENV_LOCAL_P(ep))
+        //
+        // TODO: this could properly forward the current block handler, but
+        // would require changes to gen_send_*
+        ADD_COMMENT(cb, "guard no block given");
```

Comment:
So `super` has a quirky behaviour (which I didn't know about until yesterday 😅) that it will "forward" the block it was passed onto it's super method (surprisingly to me in both the `super` and `super()` forms).

Here's how it's implemented in the interpreter https://github.com/ruby/ruby/blob/929cc615a749f467809a865a3d40adcc0b58c667/vm_args.c#L917-L919 and the test that failed without this https://github.com/ruby/ruby/blob/master/bootstraptest/test_block.rb#L584-L598

Reply:
Wow that's kind of awful.

Reply:
Yeah, super does a weird thing where super-with-no-args means pass the arguments (including the block) it was originally called with. So if you want to pass to a parent method with no args, for instance, then you need "super()" instead of "super". That's also why super isn't (and can't be) a method - it doesn't act like a method call :-/

Reply:
🤦‍♀️

Reply:
Even stranger, super-with-args ALWAYS passes the block